### PR TITLE
3.0

### PR DESCRIPTION
**Pull request by Standret**

### CODE REVIEW

File: STT/Core/Protocols.swift [hunk 1]
```diff
@@ -26,10 +26,20 @@
...
 public protocol ErrorType: Error {
     var message: ErrorMessage { get }
 }
 
+@available(swift, deprecated: 5.0)
 public protocol ServerErrorType: Decodable {
     var description: String { get }
 }
+
+/// Conform error to this protocol if yoi wan't to use custom error message
```
Comment (shamanGOD) [inline STT/Core/Protocols.swift]: Typo, yoi wan't -> you want

File: STT/Core/Protocols.swift [hunk 2]
```diff
@@ -26,10 +26,20 @@
...
+/// By default no one error conform to this
```
Comment (shamanGOD) [inline STT/Core/Protocols.swift]: Should Probably  be :
By default, errors don't conform to this

File: STT/Core/SttLog.swift [hunk 3]
```diff
@@ -27,86 +27,133 @@
...
-        formatter.locale = Locale.current
-        formatter.timeZone = TimeZone.current
-        formatter.dateFormat = "MM-dd-yyyy HH:mm:ss:SSSS"
-        
-        return formatter.string(from: value)
+    private init() { }
+
+    public func log(_ closure: () -> Any?, level: SttLogLevel, functionName: String, fileName: String, lineNumber: Int) {
+        guard destinations.isEnabledFor(level: level) else { return }
+        guard let closureResult = closure() else { return }
```
Comment (shamanGOD) [inline STT/Core/SttLog.swift]: can be in single guard statement

 `guard destinations.isEnabledFor(level: level),
  let closureResult = closure() else { return }`

File: STT/NotificationBanner/FloatingNotificationBanner.swift [hunk 4]
```diff
@@ -110,7 +111,7 @@ public class FloatingNotificationBanner: UIView, NotificationBannerType {
     
     /// Closure that will be executed if the notification banner is tapped
     public lazy var onTap: (() -> Void) = { [unowned self] in
-        let alertController = UIAlertController(title: self.lblTitle.text, message: self.lblDescription.text, preferredStyle: .alert)
+        let alertController = UIAlertController(title: self.lblTitle.text, message: (self.lblDescription.text ?? "") + "\n-----------\n" + (self.debugDescriptionText ?? ""), preferredStyle: .alert)
```
Comment (shamanGOD) [inline STT/NotificationBanner/FloatingNotificationBanner.swift]: can we use better alignment here?

**Review by shamanGOD — CHANGES_REQUESTED** [review]
few nites. Also please fix 3 and 4 way bindings

**Review by shamanGOD — APPROVED** [review]
LGTM!

File: STT/Extensions/DateExtensions.swift [hunk 5]
```diff
@@ -161,4 +172,16 @@ public extension Date {
...
+    
+    ///
+    /// Detect if time format is in 12hr or 24hr format
+    ///
+    func is12hClockFormat() -> Bool {
+        let formatString = DateFormatter.dateFormat(
+            fromTemplate: "j",
+            options: 0,
+            locale: Locale.current
+        )!
```
Comment (RomanKovalchukDev) [inline STT/Extensions/DateExtensions.swift:184]: Force unwrap 😢

**Review by RomanKovalchukDev — APPROVED** [review]

File: STT/Extensions/DateExtensions.swift (hunk 5 above)
Comment (Standret) [reply STT/Extensions/DateExtensions.swift:184]: I believe in this case it's acceptable as we have definite and static data

**Pull request merged**